lib/Sema/CSDiagnostics.cpp

```diff
@@ -2844,3 +2844,46 @@ bool ExtraneousReturnFailure::diagnoseAsError() {
   emitDiagnostic(anchor->getLoc(), diag::cannot_return_value_from_void_func);
   return true;
 }
+
+bool CollectionElementContextualFailure::diagnoseAsError() {
+  auto *anchor = getAnchor();
+
+  auto eltType = getFromType();
+  auto contextualType = getToType();
+
+  Optional<InFlightDiagnostic> diagnostic;
+  if (isa<ArrayExpr>(getRawAnchor())) {
+    diagnostic.emplace(emitDiagnostic(anchor->getLoc(),
+                                      diag::cannot_convert_array_element,
+                                      eltType, contextualType));
+  }
+
+  if (isa<DictionaryExpr>(getRawAnchor())) {
```

Comment:
This could be an `else` that asserts that `isa<DictionaryExpr>(getRawAnchor())`, because we only create this kind of failure with an array or dictionary literal anchor. Or are you intentionally being defensive here?

Reply:
Yeah, that's intention, I want to use the same fix/diagnostic to diagnose #21811